File: src/ZstShowtime/ZstShowtime/ZstNode.cs
using Newtonsoft.Json;
using NetMQ;
using NetMQ.Sockets;
using System;
using System.Net;
using System.Linq;
using System.Collections;
using System.Collections.Generic;

namespace ZST
{
    public class ZstNode
    {
        // Constants
        // ---------

        // Replies
        public static string REPLY = "zst_reply";
        public static string OK = "zst_ok";

        // Methods
        public static string PUBLISH_METHODS_CHANGED = "zst_act_announce_method_change";
        public static string REPLY_REGISTER_METHOD = "reply_register_method";
        public static string REPLY_REGISTER_NODE = "reply_register_node";
        public static string REPLY_NODE_PEERLINKS = "reply_node_peerlinks";
        public static string REPLY_METHOD_LIST = "reply_list_methods";
        public static string REPLY_ALL_PEER_METHODS = "reply_all_peer_methods";


        // Member variables
        // ----------------
        protected string m_nodeId;
        protected Dictionary<string, ZstMethod> m_methods;
        protected Dictionary<string, ZstMethod> m_internalNodeMethods;
        protected Dictionary<string, ZstPeerLink> m_peers;
        protected string m_stageAddress;
        protected string m_replyAddress;
        protected string m_publisherAddress;

        // Zmq variables
        protected NetMQContext m_ctx;
        protected ResponseSocket m_reply;
        protected PublisherSocket m_publisher;
        protected SubscriberSocket m_subscriber;
        protected RequestSocket m_stage;
        protected ZstPoller m_pollerThread;


        // Constructors
        // ------------
        public ZstNode(string nodeId)
        {
            init(nodeId, "");
        }
        public ZstNode(string nodeId, string stageAddress)
        {
            init(nodeId, stageAddress);
        }

        public void init(string nodeId, string stageAddress)
        {
            m_nodeId = nodeId;
            m_methods = new Dictionary<string, ZstMethod>();
            m_internalNodeMethods = new Dictionary<string, ZstMethod>();
            m_peers = new Dictionary<string, ZstPeerLink>();
            m_stageAddress = stageAddress;
            
            //Register internal methods to the callback dictionary
            registerInternalMethods();

            m_ctx = NetMQContext.Create();
            m_reply = m_ctx.CreateResponseSocket();
            m_publisher = m_ctx.CreatePublisherSocket();
            m_subscriber = m_ctx.CreateSubscriberSocket();

            // Binding ports
            IPHostEntry host = Dns.GetHostEntry(Dns.GetHostName());
            string hostIP = "127.0.0.1";
            foreach (IPAddress ip in host.AddressList)
            {
                if (ip.AddressFamily == System.Net.Sockets.AddressFamily.InterNetwork)
                {
                    hostIP = ip.ToString();
                    break;
                }
            }

            string address = "tcp://" + hostIP;
            int port = m_reply.BindRandomPort(address);
            m_replyAddress = address + ":" + port;

            port = m_publisher.BindRandomPort(address);
            m_publisherAddress = address + ":" + port;

            // Connect to stage
            if (!string.IsNullOrEmpty(stageAddress))
            {
                m_stage = m_ctx.CreateRequestSocket();
                m_stage.Connect(m_stageAddress);
                Console.WriteLine("Stage located at " + m_stage.Options.GetLastEndpoint);
                Console.WriteLine("Node reply on address " + m_replyAddress);
                Console.WriteLine("Node publisher on address " + m_publisherAddress);
            }

            // Subscribe to all incoming messages
            m_subscriber.SubscribeToAll();

            // Bind event listeners to sockets
            m_reply.ReceiveReady += handleReplyRequests;
            m_subscriber.ReceiveReady += receiveMethodUpdate;


            // Intialize Poller
            m_pollerThread = new ZstPoller();
            m_pollerThread.poller.AddSocket(m_subscriber);
            m_pollerThread.poller.AddSocket(m_reply);
            m_pollerThread.Start();
        }

        private void registerInternalMethods()
        {
            m_internalNodeMethods[REPLY_REGISTER_NODE] = new ZstMethod(REPLY_REGISTER_NODE, m_nodeId, ZstMethod.READ, null, replyRegisterNode);
            m_internalNodeMethods[REPLY_REGISTER_METHOD] = new ZstMethod(REPLY_REGISTER_METHOD, m_nodeId, ZstMethod.READ, null, replyRegisterMethod);
            m_internalNodeMethods[REPLY_NODE_PEERLINKS] = new ZstMethod(REPLY_NODE_PEERLINKS, m_nodeId, ZstMethod.READ, null, replyNodePeerlinks);
            m_internalNodeMethods[REPLY_METHOD_LIST] = new ZstMethod(REPLY_METHOD_LIST, m_nodeId, ZstMethod.READ, null, replyRegisterNode);
            m_internalNodeMethods[REPLY_ALL_PEER_METHODS] = new ZstMethod(REPLY_ALL_PEER_METHODS, m_nodeId, ZstMethod.READ, null, replyRegisterNode);
        }

        /// <summary>Close and dispose of all sockets/pollers/threads</summary>
        public bool cleanup()
        {
            m_pollerThread.IsDone = true;
            m_pollerThread.Update();

            m_reply.Dispose();
            m_publisher.Dispose();
            m_subscriber.Dispose();

            if (m_stage != null)
                m_stage.Dispose();

            m_ctx.Dispose();
            return true;
        }

        /// <summary>Incoming request handler</summary>
        protected void handleReplyRequests(object sender, NetMQSocketEventArgs e)
        {
            if (e.ReceiveReady)
            {
                MethodMessage msg = recv(e.Socket);
                if (m_internalNodeMethods.Keys.Contains(msg.method))
                    m_internalNodeMethods[msg.method].run(msg.data);
            }
        }

        /// <summary>Method update handler</summary>
        protected void receiveMethodUpdate(object sender, NetMQSocketEventArgs e)
        {
            if (e.ReceiveReady)
            {
                MethodMessage msg = recv(e.Socket);
                Console.WriteLine("Recieved method '" + msg.method + "' from '" + msg.data.node + "' with value '" + msg.data.output.ToString() + "'");
                if (m_methods.Keys.Contains(msg.method))
                    m_methods[msg.method].run(msg.data);
            }
        }


        // Node registration
        //------------------
        public void requestRegisterNode(){
            requestRegisterNode(m_stage);
        }

        /// <summary>Request this node to be registered on another remote node</summary>
        public void requestRegisterNode(NetMQSocket socket)
        {
            Console.WriteLine("REQ-->: Requesting remote node to register our addresses. Reply:" + m_replyAddress + ", Publisher:" + m_publisherAddress);
            
            Dictionary<string, object> requestArgs = new Dictionary<string, object>(){
                {ZstPeerLink.REPLY_ADDRESS, m_replyAddress},
                {ZstPeerLink.PUBLISHER_ADDRESS, m_publisherAddress}};
            ZstMethod request = new ZstMethod(REPLY_REGISTER_NODE, m_nodeId, "", requestArgs);

            send(socket, REPLY_REGISTER_NODE, request); 
            MethodMessage msg = recv(socket);

            if (msg.method == OK)
                Console.WriteLine("REP<--: Remote node acknowledged our addresses. Reply:" + m_replyAddress + ", Publisher:" + m_publisherAddress);
            else
                Console.WriteLine("REP<--:Remote node returned " + msg.method + " instead of " + OK);
        }

        /// <summary>Reply to another node's request for registration</summary>
        protected object replyRegisterNode(ZstMethod methodData)
        {
            if (m_peers.Keys.Contains((string)methodData.node))
                Console.WriteLine("'" + m_nodeId + "' already registered. Overwriting");

            string nodeId = (string)methodData.node;
            m_peers[nodeId] = new ZstPeerLink(
                nodeId,
                (string)methodData.args[ZstPeerLink.REPLY_ADDRESS],
                (string)methodData.args[ZstPeerLink.PUBLISHER_ADDRESS]);

            subscribeTo(m_peers[nodeId]);
            send(m_reply, OK);
            Console.WriteLine("Registered node '" + nodeId + "'. Reply:" + m_peers[nodeId].replyAddress + ", Publisher:" + m_peers[nodeId].publisherAddress);
            return null;
        }

        /// <summary> Subscribe to messages coming from an external node</summary>
        public void subscribeTo(ZstPeerLink peer)
        {
            m_subscriber.Connect(peer.publisherAddress);
            m_subscriber.SubscribeToAll();
            m_peers[peer.name] = peer;

            Console.WriteLine("Connected to peer on " + peer.publisherAddress);
        }


        //Remote method registration
        //------------------------------------
        /// <summary>Registers a local method on a remote node</summary>
        public void requestRegisterMethod(string method, string accessMode){
            requestRegisterMethod(method, accessMode, null, null, m_stage);
        }

        /// <summary>Registers a local method on a remote node</summary>
        public void requestRegisterMethod(string method, string accessMode, Dictionary<string, object> args){
            requestRegisterMethod(method, accessMode, args, null, m_stage);
        }

        /// <summary>Registers a local method on a remote node</summary>
        public void requestRegisterMethod(string method, string accessMode, string[] args){
            Dictionary<string, object> argsDict = new Dictionary<string, object>();
            foreach (string arg in args)
                argsDict[arg] = "";
            requestRegisterMethod(method, accessMode, argsDict, null, m_stage);
        }

        /// <summary>Registers a local method on a remote node</summary>
        public void requestRegisterMethod(string method, string accessMode, Dictionary<string, object> args, NetMQSocket socket){
            requestRegisterMethod(method, accessMode, args, null, socket);
        }

        /// <summary>Registers a local method on a remote node</summary>
        public void requestRegisterMethod(string method, string accessMode, Dictionary<string, object> args, Func<ZstMethod, object> callback){
            requestRegisterMethod(method, accessMode, args, callback, m_stage);
        }

        /// <summary>Registers a local method on a remote node</summary>
        public void requestRegisterMethod(string method, string accessMode, Dictionary<string, object> args, Func<ZstMethod, object> callback, NetMQSocket socket)
        {
            Console.WriteLine("REQ-->: Registering method " + method + " with remote node.");

            //Register local copy of our method first
            m_methods[method] = new ZstMethod(method, m_nodeId, accessMode, args, callback);

            //Register method copy on remote node
            send(socket, REPLY_REGISTER_METHOD, m_methods[method]);
            MethodMessage msg = recv(socket);

            if (msg.method == OK)
                Console.WriteLine("REP<--: Remote node acknowledged our method '" + method + "'");
            else
                Console.WriteLine("REP<--:Remote node returned " + msg.method + " instead of " + OK);
        }

        /// <summary>Reply to another node's method registration request</summary>
        protected object replyRegisterMethod(ZstMethod methodData)
        {
            if (m_peers[methodData.node].methods.Keys.Contains(methodData.name))
                Console.WriteLine("'" + methodData.name + "' already registered on node " + methodData.node + ". Overwriting");

            ZstMethod localMethod = new ZstMethod(
                methodData.name,
                methodData.node,
                methodData.accessMode,
                methodData.args
            );

            m_peers[methodData.node].methods[methodData.name] = localMethod;

            send(m_reply, OK);
            Console.WriteLine("Registered method '" + localMethod.name + "'. Origin:" + localMethod.node + ", AccessMode:" + localMethod.accessMode + ", Args:" + localMethod.args);
            return null;
        }


        // Node peerlink accessors
        //------------------------
        /// <summary>Request a dictionary of peers nodes linked to the remote node</summary>
        public Dictionary<string, ZstPeerLink> requestNodePeerlinks()
        {
            return requestNodePeerlinks(m_stage);
        }

        /// <summary>Request a dictionary of peers nodes linked to the remote node</summary>
        public Dictionary<string, ZstPeerLink> requestNodePeerlinks(NetMQSocket socket)
        {
            send(socket, REPLY_NODE_PEERLINKS);
            Dictionary<string, object> peerDict = JsonConvert.DeserializeObject<Dictionary<string, object>>(recv(socket).data.output.ToString());
            return ZstPeerLink.buildLocalPeerlinks(peerDict);
        }

        /// <summary>Reply to another node's request for this node's linked peers</summary>
        protected object replyNodePeerlinks(ZstMethod methodData)
        {
            Dictionary<string, object> peerDict = new Dictionary<string, object>();
            foreach (KeyValuePair<string, ZstPeerLink> peer in m_peers)
                peerDict[peer.Key] = peer.Value.as_dict();
            ZstMethod request = new ZstMethod(REPLY_NODE_PEERLINKS, m_nodeId, "", null);
            request.output = peerDict;
            send(m_reply, OK, request);
            return null;
        }

        public Dictionary<string, ZstMethod> requestMethodList()
        {
            return requestMethodList(m_stage);
        }


        // Node method accessors
        //----------------------
        /// <summary>Request a dictionary of methods on a remote node</summary>
        public Dictionary<string, ZstMethod> requestMethodList(NetMQSocket socket)
        {
            send(socket, REPLY_METHOD_LIST);
            return ZstMethod.buildLocalMethods((Dictionary<string, object>)recv(socket).data.output);
        }

        /// <summary>Reply with a list of methods this node owns</summary>
        public object replyMethodList(ZstMethod methodData)
        {
            Dictionary<string, object> methodDict = new Dictionary<string, object>();
            foreach (KeyValuePair<string, ZstMethod> method in m_methods)
                methodDict[method.Key] = method.Value.as_dict();
            ZstMethod request = new ZstMethod(REPLY_NODE_PEERLINKS, m_nodeId, "", null);
            request.output = methodDict;
            send(m_reply, OK, request);
            return null;
        }


        // Get all methods on all connected peers
        //----------------------
        /// <summary>Request a dictionary of methods on a remote node</summary>
        public Dictionary<string, ZstMethod> requestAllPeerMethods()
        {
            return requestAllPeerMethods(m_stage);
        }

        /// <summary>Request a list of all available methods provided by all connected peers on the remote node</summary>
        public Dictionary<string, ZstMethod> requestAllPeerMethods(NetMQSocket socket)
        {
            send(socket, REPLY_METHOD_LIST);
            return ZstMethod.buildLocalMethods((Dictionary<string, object>)recv(socket).data.output);
        }

        /// <summary>Reply with a list of all available methods provided by all connected peers on the remote node</summary>
        public object replyAllPeerMethods(ZstMethod methodData)
        {
            Dictionary<string, object> methodDict = new Dictionary<string, object>();
            foreach (KeyValuePair<string, ZstPeerLink> peer in m_peers)
            {
                foreach (KeyValuePair<string, ZstMethod> method in peer.Value.methods)
                    methodDict[method.Key] = method.Value.as_dict();
            }
            ZstMethod request = new ZstMethod(REPLY_NODE_PEERLINKS, m_nodeId, "", null);
            request.output = methodDict;
            send(m_reply, OK, request);
            return null;
        }


        // Remote send/recieve methods
        // ---------------------------
        /// <summary>Send a message to a remote node</summary>
        protected static void send(NetMQSocket socket, string method)
        {
            send(socket, method, null);
        }

        /// <summary>Send a message to a remote node using method info</summary>
        protected static void send(NetMQSocket socket, string method, ZstMethod methodData)
        {
            NetMQMessage message = new NetMQMessage();
            message.Append(method);
            if (methodData != null)
            {
                string data = JsonConvert.SerializeObject(methodData.as_dict());
                JsonConvert.SerializeObject(data, Formatting.Indented);
                message.Append(data);
            }
            else
            {
                message.Append("{}");
            }
            socket.SendMessage(message);
        }

        
        // Method publishing / controlling
        // -------------------------------
        /// <summary>Updates the value of a local variable by name</summary>
        public void updateLocalMethodByName(string method, object methodvalue)
        {
            if (m_methods.Keys.Contains(method))
                updateLocalMethod(m_methods[method], methodvalue);
        }

        /// <summary>Updates the value of a local variable</summary>
        public void updateLocalMethod(ZstMethod method, object value)
        {
            if (method.node == m_nodeId)
            {
                method.output = value;
                send(m_publisher, method.name, method);
            }
        }

        /// <summary>Triggers a remote method with arguments</summary>
        public void updateRemoteMethod(ZstMethod method, Dictionary<string, object> args)
        {
            if (m_peers.Keys.Contains(method.node))
            {
                if (ZstMethod.compareArgLists(args, m_peers[method.node].methods[method.name].args))
                {
                    ZstMethod methodRequest = new ZstMethod(method.name, m_nodeId, method.accessMode, args);
                    send(m_publisher, method.name, methodRequest);
                }
            }
        }

        /// <summary>Receive a message from a remote node</summary>
        protected static MethodMessage recv(NetMQSocket socket)
        {
            return recv(socket, false);
        }

        /// <summary>Receive a message from a remote node</summary>
        protected static MethodMessage recv(NetMQSocket socket, bool dontWait)
        {
            try
            {
                NetMQMessage message = socket.ReceiveMessage(dontWait);
                string method = message[0].ConvertToString();

                //Dictionary<string, object> data = new Dictionary<string, object>();
                string jsonStr = (message[1].ConvertToString());
                Dictionary<string, object> data = JsonConvert.DeserializeObject<Dictionary<string, object>>(jsonStr);

                return new MethodMessage(method, ZstMethod.dictToZstMethod(data));
            }
            catch (NetMQException e)
            {

            }
            return new MethodMessage("", null);
        }
    }



    /// <summary>Struct to hold message information</summary>
    public struct MethodMessage
    {
        public string method;
        public ZstMethod data;
        public MethodMessage(string methodName, ZstMethod methodData)
        {
            method = methodName;
            data = methodData;
        }
    }
}

File: python/Showtime/zst_node.py
import sys
import json
import time
import socket
from zst_base import *
from zst_socket import ZstSocket
from zst_method import ZstMethod
from zst_peerLink import ZstPeerLink


class ZstNode(ZstBase):

    # Message constants
    REPLY = "zst_reply"
    OK = "zst_ok"
    ERR = "zst_err"

    # Event action types
    PUBLISH_METHODS_CHANGED = "zst_act_announce_method_change"
    REPLY_REGISTER_METHOD = "reply_register_method"
    REPLY_REGISTER_NODE = "reply_register_node"
    REPLY_REGISTER_INCOMING = "reply_register_incoming"
    REPLY_NODE_PEERLINKS = "reply_node_peerlinks"
    REPLY_METHOD_LIST = "reply_list_methods"
    REPLY_ALL_PEER_METHODS = "reply_all_peer_methods"
    DISCONNECT_PEER = "disconnect_peer"

    def __init__(self, nodeId, stageAddress=None):
        ZstBase.__init__(self, name=nodeId)

        self.methods = {}
        self.peers = {}

        # Sockets
        self.stageAddress = stageAddress
        if self.stageAddress:
            self.stageAddress = "tcp://" + self.stageAddress
        self.replyAddress = None
        self.publisherAddress = None

        self.reply = ZstSocket(self.ctx, zmq.REP, self.incomingQueue, "reply")
        self.publisher = ZstSocket(self.ctx, zmq.PUB, self.incomingQueue, "publisher")
        self.subscriber = ZstSocket(self.ctx, zmq.SUB, self.incomingQueue, "subscriber")
        self.stage = ZstSocket(self.ctx, zmq.REQ, self.incomingQueue, "stage") if stageAddress else None

        # Binding ports
        address = 'tcp://{0}:*'.format(socket.gethostbyname(socket.gethostname()))
        self.reply.socket.bind(address)
        self.replyAddress = self.reply.socket.getsockopt(zmq.LAST_ENDPOINT)

        self.publisher.socket.bind(address)
        self.publisherAddress = self.publisher.socket.getsockopt(zmq.LAST_ENDPOINT)

        # Subscribe to all incoming messages
        self.subscriber.socket.setsockopt(zmq.SUBSCRIBE, "")

    def run(self):
         # Connect to stage
        if self.stageAddress:
            self.stage.socket.connect(self.stageAddress)

        ZstBase.run(self)

    def close(self):
        print "PUB-->: Announcing that we're leaving."  
        self.publisher.send(ZstNode.DISCONNECT_PEER, ZstMethod(ZstNode.DISCONNECT_PEER, self.id))

        for peer, peerlink in self.peers.iteritems():
            peerlink.disconnect()

        if self.stage:
            self.stage.stop()

        self.reply.stop()
        self.publisher.stop()
        self.subscriber.stop()

        if self.stage:
            print "stage:{0}".format(self.stage.is_alive())
        print "reply:{0}".format(self.reply.is_alive())
        print "publisher:{0}".format(self.publisher.is_alive())
        print "subscriber:{0}".format(self.subscriber.is_alive())
        ZstBase.close(self)

    def disconnect_peer(self, methodData):
        print "Peer '{0}' is leaving.".format(methodData.node)
        if methodData.node in self.peers:
            print "Disconnecting address {0}".format(self.peers[methodData.node].publisherAddress)
            self.subscriber.socket.disconnect(self.peers[methodData.node].publisherAddress)
            self.peers[methodData.node].disconnect()
            del self.peers[methodData.node]
        print "Peers now contains:"
        for peer in self.peers.iteritems():
            print peer

    # ----------------------------------------------
    # Recieve updates from nodes we're interested in
    # ----------------------------------------------
    def receive_message(self, message):
        if message.method in self.methods:
            print "Matched local method '{0}' from '{1}' with value '{2} and args {3}'".format(
                message.method,
                message.data.node,
                message.data.output,
                message.data.args)
            self.methods[message.method].run(message.data)

        if hasattr(self, message.method):
            print "Matched internal method '{0}'".format(message.method)
            getattr(self, message.method)(message.data)

        if isinstance(message.data, ZstMethod):
            if message.data.node in self.peers:
                if message.method in self.peers[message.data.node].methods:
                    if self.peers[message.data.node].methods[message.method].callback:
                        self.peers[message.data.node].methods[message.method].callback(message.data)

    # ------------------------------------------------------
    # Request/Reply this node be registered to a remote peer
    # ------------------------------------------------------
    def request_register_node(self, nodesocket=None):
        socket = nodesocket if nodesocket else self.stage

        # Request a port for this equipment to bind as its outgoing sender
        print "REQ-->: Requesting remote node to register our addresses. Reply:{0}, Publisher:{1}".format(self.replyAddress, self.publisherAddress)
        request = ZstMethod(
            name=ZstNode.REPLY_REGISTER_NODE,
            node=self.id,
            args={
                ZstPeerLink.REPLY_ADDRESS: self.replyAddress,
                ZstPeerLink.PUBLISHER_ADDRESS: self.publisherAddress
            })
        socket.send_immediate(ZstNode.REPLY_REGISTER_NODE, request)
        message = socket.recv_immediate()
        if message:
            if message.method == ZstNode.OK:
                print "REP<--: Remote node acknowledged our addresses. Reply:{0}, Publisher:{1}".format(self.replyAddress, self.publisherAddress)
                return True
            else:
                print "REP<--:Remote node returned {0} instead of {1}.".format(message.method, ZstNode.OK)
            return False
        else:
            print "No response from node"

    def reply_register_node(self, methodData):
        if methodData.node in self.peers:
            print "'{0}' node already registered. Overwriting".format(methodData.node)

        self.peers[methodData.node] = ZstPeerLink(
            methodData.node,
            methodData.args[ZstPeerLink.REPLY_ADDRESS],
            methodData.args[ZstPeerLink.PUBLISHER_ADDRESS])

        self.subscribe_to(self.peers[methodData.node])

        self.reply.send_immediate(ZstNode.OK)
        print "Registered node '{0}'. Reply:{1}, Publisher:{2}".format(methodData.node, self.peers[methodData.node].replyAddress, self.peers[methodData.node].publisherAddress)

    # -------------------------------------------
    # Subscribe to messages from an external node
    # -------------------------------------------
    def subscribe_to(self, peerlink):
        # Connect to peer
        self.subscriber.socket.connect(peerlink.publisherAddress)
        self.subscriber.socket.setsockopt(zmq.SUBSCRIBE, '')
        self.peers[peerlink.name] = peerlink

        # Register with the event poller
        print "Connected to peer on", self.subscriber.socket.getsockopt(zmq.LAST_ENDPOINT)

    def connect_to_peer(self, peer):
        request = ZstSocket(self.ctx, zmq.REQ, self.incomingQueue, "request")
        request.socket.connect(peer.replyAddress)
        if self.request_register_node(request):
            if not peer.name in self.peers:
                self.peers[peer.name] = peer
            self.peers[peer.name].request = request
            print "Connected to peer " + peer.name

    # ---------------------------------------
    # Request/reply a method to be registered
    # ---------------------------------------
    def request_register_method(self, method, mode, args=None, callback=None, nodesocket=None):
        socket = nodesocket if nodesocket else self.stage
        args = args if args else {}
        print "REQ-->: Registering method '%s' with remote node." % method

        # Register local copy of our method first
        self.methods[method] = ZstMethod(method, self.id, mode, args, callback)

        # Make remote copy of local method on stage
        socket.send_immediate(ZstNode.REPLY_REGISTER_METHOD, self.methods[method])
        message = socket.recv_immediate()

        if message.method == ZstNode.OK:
            print "REP<--: Remote node acknowledged our method '{0}'".format(method)
        else:
            print "REP<--: Remote node returned {0} instead of {1}.".format(message.method, ZstNode.OK)

        return self.methods[method]

    def reply_register_method(self, methodData):
        if methodData.name in self.methods:
            print "'{0}' method already registered. Overwriting".format(methodData.name)

        self.peers[methodData.node].methods[methodData.name] = methodData
        self.reply.send_immediate(ZstNode.OK)
        print "Registered method '{0}'. Origin:{1}, AccessMode:{2}, Args:{3}".format(methodData.name, methodData.node, methodData.accessMode, methodData.args)

    # -----------------------------------------------------------------
    # Request/Reply a list of all available nodes connected to this one
    # -----------------------------------------------------------------
    def request_node_peerlinks(self, nodesocket=None):
        socket = nodesocket if nodesocket else self.stage
        socket.send_immediate(ZstNode.REPLY_NODE_PEERLINKS)
        message = socket.recv_immediate()
        return ZstPeerLink.build_local_peerlinks(message.data.output)

    def reply_node_peerlinks(self, args):
        peerList = {}
        for peerName, peerData in self.peers.iteritems():
            peerList[peerName] = peerData.as_dict()

        request = ZstMethod(
            name=ZstNode.REPLY_NODE_PEERLINKS,
            node=self.id,
            output=peerList)
        self.reply.send_immediate(ZstNode.OK, request)

    # -------------------------------------------------------------------
    # Request/Reply a list of all available methods provided by this node
    # -------------------------------------------------------------------
    def request_method_list(self, nodesocket=None):
        socket = nodesocket if nodesocket else self.stage
        socket.send_immediate(ZstNode.REPLY_METHOD_LIST)
        message = socket.recv_immediate()
        return ZstMethod.build_local_methods(message.data.output)

    def reply_method_list(self, args):
        methodList = {}
        for name, method in self.methods.iteritems():
            methodList[name] = method.as_dict()

        request = ZstMethod(
            name=ZstNode.REPLY_METHOD_LIST,
            node=self.nodeId,
            output=methodList)
        self.reply.send_immediate(ZstNode.OK, request)

    # -------------------------------------------------------------
    # Request/Reply a list of all available methods provided by all
    # connected peers to the remote node
    # -------------------------------------------------------------
    def request_all_peer_methods(self, nodesocket=None):
        socket = nodesocket if nodesocket else self.stage
        socket.send_immediate(ZstNode.REPLY_ALL_PEER_METHODS)
        message = recv_immediate(socket)
        return ZstMethod.build_local_methods(message.methodData.output)

    def reply_all_peer_methods(self, args):
        methodList = {}
        for peerName, peer in self.peers.iteritems():
            for methodName, method in peer.methods.iteritems():
                methodList[methodName] = method.as_dict()
        request = ZstMethod(
            name=REPLY_ALL_PEER_METHODS,
            node=self.nodeId,
            output=methodList)
        self.reply.send_immediate(ZstNode.OK, request)

    # ----------------------------------------------------------
    # Publish updates from a node method to all interested nodes
    # ----------------------------------------------------------
    def update_local_method_by_name(self, methodname, methodvalue):
        method = self.methods[methodname]
        return self.update_local_method(method, methodvalue)

    def update_local_method(self, method, methodvalue):
        # Only update local values on methods WE own
        if method.node == self.id:
            method.output = methodvalue
            if method.accessMode == ZstMethod.RESPONDER:
                print "Sending on reply..."
                self.reply.send_immediate(method.name, method)
            else:
                self.publisher.send(method.name, method)
        else:
            print "We don't own this method!"
        return methodvalue

    def update_remote_method_by_name(self, methodname, methodargs=None):
        method = self.methods[methodname]
        return self.update_remote_method(method, methodargs)

    def update_remote_method(self, method, methodargs=None):
        socket = self.peers[method.node].request if method.accessMode == ZstMethod.RESPONDER else self.publisher
        if method.node in self.peers:
            methodDict = method.as_dict()
            if ZstMethod.compare_arg_lists(methodDict[ZstMethod.METHOD_ARGS], methodargs):
                methodData = ZstMethod(
                    name=method.name,
                    node=method.node,
                    args=methodargs)
                socket.send_immediate(method.name, methodData)
                if method.accessMode == ZstMethod.RESPONDER:
                    return socket.recv_immediate().data
            else:
                print "Mismatch on arguments being sent to remote node!"
                raise
        else:
            print "Method destination node not in connected peers list!"
        return None

    def subscribe_to_method(self, method, callback):
        if method.node in self.peers:
            if method.name in self.peers[method.node].methods:
                self.peers[method.node].methods[method.name].callback = callback


# Test cases
if __name__ == '__main__':

    if len(sys.argv) > 2:
        node = ZstNode(sys.argv[1], sys.argv[2])
        node.start()
        nodeList = node.request_node_peerlinks()

        try:
            while True:
                time.sleep(1)
        except KeyboardInterrupt:
            node.close()
            print "Finished"

    else:
        print "Please provide a name for this node!"
